### PR TITLE
Dashboard

### PR DESCRIPTION
**Pull request by erikstmartin**
Refactoring of InstanceMonitor to use a single notification channel and a callback method to keep the exposed interface from having to know anything about the InstanceListener object, and monitoring many channels.

Also refactored instancesocket.go to support cleaning up after timeouts, and stale web sockets that failed to read/write, this refactoring also includes modifications so that it uses the new InstanceMonitor implementation

Several changes to the logging inside dashboard that were merged from andrey's branch

### CODE REVIEW

**Comment by skelterjohn** [comment]
I get what you were saying about factoring out the listen channels.

I'm not sure that's a good idea. In many cases (for instance, the ServiceClient) the adds/removes have to come in on channels so they can be part of a mux()'s select{} statement. If only callbacks are allowed, I would simply create a callback that sent on a channel.

Maybe make an InstanceListener.SetCallback(aCallbackFunction) that would intercept the things coming in on the channel, if the user definitely wanted a callback rather than a channel? And leave the channel exported so it can be selected on directly?

**Comment by skelterjohn** [comment]
also, gofmt :)

**Comment by erikstmartin** [comment]
That's a valid point, so we really have 2 options here.

1) we can expose the notificationChannel so that you can utilize it in mux() and make the callback optional, and can be done in a .SetCallback method as mentioned above.

2) you could pass in a callback method that sends the notification down your own channel that is being selected on in mux()

which do we prefer?

**Comment by skelterjohn** [comment]
I would prefer method 1, since method 2 involves an extra goroutine. Goroutines are light, but they add up, especially when looking through stack traces.

**Comment by erikstmartin** [comment]
Ok guys, all suggested changes have been made. Looking for a thumbs up :)

**Comment by mirtchovski** [comment]
LGTM

File: client/instancemonitor.go [hunk 1]
```diff
@@ -154,13 +179,12 @@ func (im *InstanceMonitor) monitorInstances() {
...
-		monitor:    im,
-		id:         id,
-		Instances:  make(map[string]service.Service),
-		doneChan:   make(chan bool),
+		query:            q,
+		monitor:          im,
+		id:               id,
+		Instances:        make(map[string]service.Service),
+		doneChan:         make(chan bool),
+		NotificationChan: make(chan InstanceListenerNotification),
```
Comment (skelterjohn) [inline client/instancemonitor.go]: This channel should be buffered, so the instance monitor doesn't have to wait for the listener to read from the channel before continuing.

Actually, this opens up a vulnerability (not a security vulnerability, but a deadlock vulnerability). If the person listening to instance notifications stops listening without informing the monitor, a deadlock will occur once the buffer on that channel is filled.

My recommendation would be to have a channel with a decent-sized buffer (perhaps 10), and a select{} with a default case to send on that channel. The default case will be empty, and the normal case will be the send you're already doing. That way if the buffer gets full and the listener is behaving poorly, future notifications will just be dropped.

The down side is that if the listener is just blocked on something that takes a few seconds, and during those few seconds a large number of instances are discovered, then the listener will not be notified about all of them.

Idea: in the default case, spawn a new goroutine whose job it is to send to the channel. When this happens, the number of goroutines goes up, and notification order gets lost, but notification order isn't important in the first place, and the #goroutines only goes up in pathalogical situations.

**Comment by erikstmartin** [comment]
order may be important if it receives an add then a delete of an instance, if you get the add second you will have an instance that isn't really live

**Comment by erikstmartin** [comment]
we also have the option of just sending a whole list again if we get too many notifications queued?

**Comment by skelterjohn** [comment]
I think it might be reasonable to just drop notifications that overflow the buffer, then, and leave it up to the client code to pull the whole list.

**Comment by mirtchovski** [comment]
> I think it might be reasonable to just drop notifications that overflow
> the buffer, then, and leave it up to the client code to pull the whole list.
> 
> you saw the timings and sizes for fully buffered (find().All()) and
> unbuffered (find().Iter().Next() in a loop) database reads -- due to
> channel copies the memory consumed by the former was about twice the size
> of the input read (100mb footprint for 50mb data). timing-wise they weren't
> significantly different, so i guess your only worry is whether to make the
> buffer static or let it grow a-la bufio.Buffer. it will get cleaned up on
> client timeout, right?

i guess the real question is how much data do you expect to accumulate in
15 seconds? for mgo simple queries i can read 50mb in 40 seconds, so
18.75mb in 15 seconds at full-throttle.

**Comment by mirtchovski** [comment]
> so i guess your only worry is whether to make the
> buffer static or let it grow a-la bufio.Buffer. it will get cleaned up on
> client timeout, right?

scratch that: there could exist a very slow reader that fits just
inside the heartbeat yet doesn't consume enough, allowing your
unconstrained buffer to always grow given a fast producer....

my bad.

**Comment by skelterjohn** [comment]
I have a solution: it's a bit funky seeming at first but it's a great
concept I've played around with on go.uik.

A stacked channel is one with a buffer of size one, where sends never
block. If the buffer is full, the thing you try to send gets "stacked" with
whatever is already there. Stacking in this context just means combining in
some relevant way.

In this case, the thing on the channel is a list of the updates to the
instance list. Some adds, some removes. Stacking an add on top would either
put the instance into the add list, or take it out of the remove list
(depending on if it were in the remove list). Stacking a remove does the
same but in reverse.

The way to stack a value is like so:

func stack(ch chan Whatever, w Whatever) {
  select {
    case oldW := <- ch:
      ch <- myStackFunction(w, oldW)
    case ch <- w:
  }
}

If there is nothing on the chan, w is put there like normal. If something
is there, it's taken off, combined with w, and then put back on.

For our particular case, it makes sense to do a stacked channel like this.

**Comment by skelterjohn** [comment]
Minor error - can't just send the stacked result on the channel. You need to select again or this isn't threads ate. Easy fix.

**Comment by erikstmartin** [comment]
I think i'm confused what you mean, so pass a list of notifications on the channel rather than a notification itself?

**Comment by skelterjohn** [comment]
it'd be a chan struct { Additions, Removals []*Service }. The first time
the client recvs from the channel, Additions would be the set of all
currently known instances and Removals would be empty. Every subsequent
recv would get the delta.

type Notification struct {
  Additions, Removals []*Service
}

type NotificationChan chan Notification

func (ch NotificationChan) Stack(n Notification) {
  for {
    select {
      case ch <- n:
        return
      case on := <- ch:
        n = union(n, on)
    }
  }
}

and the union function is the one I described before that turns two ordered
Notification instances and turns them into a single combined Notification.
Since we're dealing with insertions to and deletions from a set, it's easy
to write this union function.

So when the InstanceMonitor gets a new service that is accepted by a query,
it creates a Notification with just that one addition and stacks it onto
the NotificationChan. If the client recvs from the channel right away,
that's great. If not, and another new service comes, the stacking makes a
Notification with the two additions, and that will be ready to be received
by the client. etc.

**Comment by skelterjohn** [comment]
LGTM

**Pull request merged**